sources/academy/webscraping/scraping_basics_python/06_locating_elements.md

```diff
@@ -0,0 +1,295 @@
+
+```py
+import httpx
+from bs4 import BeautifulSoup
+
+url = "https://warehouse-theme-metal.myshopify.com/collections/sales"
+response = httpx.get(url)
+response.raise_for_status()
+
+html_code = response.text
+soup = BeautifulSoup(html_code, "html.parser")
+
+for product in soup.select(".product-item"):
+    title = product.select_one(".product-item__title").text
+    price = product.select_one(".price").text
+    print(title, price)
+```
+
+This program does the same as the one we already had, but its code is more concise.
+
+## Precisely locating price
+
+In the output we can see that the price isn't located precisely. For each product, our scraper also prints the text `Sale price`. Let's look at the HTML structure again. Each bit containing the price looks like this:
+
+```html
+<span class="price">
+  <span class="visually-hidden">Sale price</span>
+  $74.95
+</span>
+```
+
+When translated to a tree of Python objects, the element with class `price` will contain several _nodes_:
+
+- Textual node with white space,
+- a `span` HTML element,
+- a textual node representing the actual amount and possibly also white space.
+
+We can use Beautiful Soup's `.contents` property to access individual nodes. It returns a list of nodes like this:
+
+```py
+["\n", <span class="visually-hidden">Sale price</span>, "$74.95"]
+```
+
+It seems like we can read the last element to get the actual amount from a list like the above. Let's fix our program:
+
+```py
+import httpx
+from bs4 import BeautifulSoup
+
+url = "https://warehouse-theme-metal.myshopify.com/collections/sales"
+response = httpx.get(url)
+response.raise_for_status()
+
+html_code = response.text
+soup = BeautifulSoup(html_code, "html.parser")
+
+for product in soup.select(".product-item"):
+    title = product.select_one(".product-item__title").text
+    price = product.select_one(".price").contents[-1]
+    print(title, price)
+```
+
+If we run the scraper now, it should print prices as only amounts:
+
+```text
+$ python main.py
+JBL Flip 4 Waterproof Portable Bluetooth Speaker $74.95
+Sony XBR-950G BRAVIA 4K HDR Ultra HD TV From $1,398.00
+Sony SACS9 10" Active Subwoofer $158.00
+Sony PS-HX500 Hi-Res USB Turntable $398.00
+...
+```
+
+Great! We have managed to use CSS selectors and walk the HTML tree to get a list of product titles and prices. But wait a second—what's `From $1,398.00`? One does not simply scrape a price! We'll need to clean that. But that's a job for the next lesson, which is about extracting data.
+
+---
+
+## Exercises
+
+These challenges are here to help you test what you’ve learned in this lesson. Try to resist the urge to peek at the solutions right away. Remember, the best learning happens when you dive in and do it yourself!
```

Comment:
Since we're always using real websites in the exercises, and we're always repeating the same intro text, it might make sense to also add some warning about selectors changing and the real solution being different from what's listed here?

Reply:
That's a great idea 👍

Reply:
If we are reusing the same structure to start this section we could remake it into a [partial](https://docusaurus.io/docs/markdown-features/react#importing-markdown) for ease of maintaining it from one place.

Reply:
Cool, I didn't realize we can have those. I'll use a partial.

Reply:
Made the partial in f775e4dbb715ae435eb510da4b2fe7dee2746853 and added a warning about the real world in 6258e5ba7c85ec85372cab79d33829efee5de48f